File: quizzes1/sixth_problem.js
const fs = require("fs");

const staffs = JSON.parse(fs.readFileSync("input_data/staff.json", "utf8"));

const findTopRatedStaffsWithChained = (staffs) => {
  return staffs.flatMap((staff) => {
    if (staff?.result?.rating <= 4) return [];

    return {
      id: staff.id,
      full_name: staff.full_name,
      email: staff.email,
      rating: staff.result.rating,
    };
  });
};

const findTopRatedStaffsWithLoop = (staffs) => {
  const topRatedStaffs = [];

  for (let i = 0; i < staffs.length; i++) {
    const staff = staffs[i];

    if (staff?.result?.rating > 4) {
      topRatedStaffs.push({
        id: staff.id,
        full_name: staff.full_name,
        email: staff.email,
        rating: staff.result.rating,
      });
    }
  }

  return topRatedStaffs;
};

console.log(
  JSON.stringify({ data: findTopRatedStaffsWithChained(staffs) }, null, 2),
);
console.log(
  JSON.stringify({ data: findTopRatedStaffsWithLoop(staffs) }, null, 2),
);

File: binV3/sixth_problem.rb
require 'json'

staffs = JSON.parse(File.read('input_data/staff.json'))

high_rated_data = staffs.select { |staff| staff['result']['rating'] > 4 }

filtered_data = high_rated_data.map do |data|
  {
    id: data['id'],
    full_name: data['full_name'],
    email: data['email'],
    rating: data['result']['rating']
  }
end

print JSON.pretty_generate(filtered_data)
